### PR TITLE
Julia format GH workflow, format all files

### PR DESCRIPTION
**Pull request by arlowhite**
Adds [julia-format](https://github.com/julia-actions/julia-format) GitHub workflow action.
Add `.JuliaFormatter.toml`
Update `development_setup.md`
format all files.

Also see [Notion notes](https://www.notion.so/Julia-Linter-Formatter-e12b052b5bef4796a888badeafc24050?pvs=4)

## Developer Setup Change

Recommendation is to uninstall the Julia Formatter plugin (or disable for your ADRIA.jl workspace if you use it in other projects). The primary VSCode Julia plugin uses `.JuliaFormatter.toml`, making the plugin redundant; the plugin only uses its own config and ignores `.JuliaFormatter.toml`. So having both installed/active adds confusion with around which formatting config is being applied.

## Formatter Config Changes

Changes from what we were using before with the formatter plugin. These settings reduce formatter changes:
* short_to_long_function_def=false
* conditional_to_if=false
* always_for_in="nothing"

Note: `always_for_in` config is strange. `false` seems to be always `=`, `true` is `in` (or value of `for_in_replacement`). `nothing` makes no change.

### CODE REVIEW

File: docs/make.jl
Comment (arlowhite) [inline docs/make.jl:6]: This is weird, not sure why the semicolon on opening parenthesis.

File: ext/AvizExt/AvizExt.jl
Comment (arlowhite) [inline ext/AvizExt/AvizExt.jl:69]: comma converted to semicolon?

**Review by arlowhite — COMMENTED** [review]
seems like there's a rule that a semicolon always goes before named args. in some cases this seems awkward to me.

File: ext/AvizExt/AvizExt.jl
Comment (ConnectedSystems) [reply ext/AvizExt/AvizExt.jl:69]: Is there a named argument afterwards? If so, that's why

File: docs/make.jl
Comment (ConnectedSystems) [reply docs/make.jl:6]: If the only arguments are keyword arguments, this is clear to me:

```julia
function blah(;
    foo=100
)
```

And is preferred over some other weirdness

```julia
function blah(
    ; foo=100
)
```

File: ext/AvizExt/AvizExt.jl
Comment (arlowhite) [reply ext/AvizExt/AvizExt.jl:69]: yes, there is, it just looks funny to me when it does `foo(; x = 1)` but I'm not seeing any option to omit the leading semicolon if all args are kwargs.

**Comment by ConnectedSystems** [comment]
> seems like there's a rule that a semicolon always goes before named args. in some cases this seems awkward to me.

Yes, this is Julia syntax. Keyword arguments are specified after a semi-colon.

File: ext/AvizExt/AvizExt.jl
Comment (ConnectedSystems) [reply ext/AvizExt/AvizExt.jl:69]: It'll be a syntax error without the semi-colon, won't it? At least I thought it would be...

File: ODE_preferences.jl
Comment (arlowhite) [inline ODE_preferences.jl:6]: For example, this could be a place where we disable formatting so they're always one line. 
```
#! format: off
code
#! format: on
```

**Comment by arlowhite** [comment]
I was searching to see if there's a way to preserve git blame history. The best way seems to be `--ignore-revs-file`
https://medium.com/codex/how-to-introduce-a-code-formatter-without-messing-up-git-history-4a16bd074c10
So IDE git blame features would ignore this commit.

File: ext/AvizExt/AvizExt.jl
Comment (arlowhite) [inline ext/AvizExt/AvizExt.jl:563]: would you prefer this instead?
```julia
wait(gl_screen)
return
```

File: ODE_preferences.jl
Comment (arlowhite) [reply ODE_preferences.jl:6]: though in this particular case, better to extract the variable.

Comment (ConnectedSystems) [reply ODE_preferences.jl]: Don't stress too much about this particular file, it'll be removed after the new model gets integrated (@Zapiano) as we won't need the package it is configuring

File: ext/AvizExt/AvizExt.jl
Comment (ConnectedSystems) [reply ext/AvizExt/AvizExt.jl:563]: Explicit return is better imo.

And yes, what you suggest is fine with me

**Comment by Zapiano** [comment]
Since we have a very critical/delicate PR, almost ready to merge, changing more than 20 files, I think say we should wait for that one to be merged before this.

**Comment by ConnectedSystems** [comment]
@arlowhite @Zapiano is the blocker for this PR now resolved?

**Comment by Zapiano** [comment]
> @arlowhite @Zapiano is the blocker for this PR now resolved?

My understanding we decided to wait until #760 is merged to main.

**Comment by arlowhite** [comment]
In our last meeting, we discussed getting this in sooner than later and allowing PRs to merge with this check failing if necessary. I've rebased and reformatted.

**Comment by arlowhite** [comment]
@Zapiano do you still want me to leave `src/scenario.jl` untouched?
or are you ok with these formatting changes?
https://github.com/open-AIMS/ADRIA.jl/pull/790/files#diff-bd5b8c618631550cda060d53f80beee397a2975f98de66d5a123ed45199ed86c

**Comment by Zapiano** [comment]
> @Zapiano do you still want me to leave `src/scenario.jl` untouched? or are you ok with these formatting changes? https://github.com/open-AIMS/ADRIA.jl/pull/790/files#diff-bd5b8c618631550cda060d53f80beee397a2975f98de66d5a123ed45199ed86c

I am going to merge #817 today, after that you can format the entire project and I review it. Is that ok?

File: ext/AvizExt/viz/taxa_dynamics.jl
Comment (arlowhite) [inline ext/AvizExt/viz/taxa_dynamics.jl:205]: not sure I like it changing if/else to elvis operator. what are your thoughts?

Comment (ConnectedSystems) [reply ext/AvizExt/viz/taxa_dynamics.jl:205]: I think It should leave these alone. If I use the ternary operator, it's for a reason, and vice versa.

**Comment by arlowhite** [comment]
What's your preference here?
returning assignment expressions seems a bit weird to me:
`return data .= (data .- mi) ./ (ma - mi)`
so, I'm thinking this is better:
```
data .= (data .- mi) ./ (ma - mi)
return data
```

**Comment by ConnectedSystems** [comment]
Agree with your thinking

File: src/analysis/rule_extraction.jl
Comment (arlowhite) [inline src/analysis/rule_extraction.jl:95]: `function` expansion. if you dislike this, I can look into options to prevent it.

Comment (ConnectedSystems) [reply src/analysis/rule_extraction.jl:95]: It's definitely more verbose than I would prefer but I guess it's okay...

In fact, probably a sign that the function should be extracted, as currently it gets recompiled every call to `print_rules()`

**Comment by ConnectedSystems** [comment]
Not a fan of these forced replacement of `in` to `∈` but I guess it's okay...

```julia
for r ∈ 1:n_rows
        for c ∈ 1:n_cols
```

File: src/analysis/rule_extraction.jl
Comment (ConnectedSystems) [reply src/analysis/rule_extraction.jl:95]: (that wasn't to say we should refactor it in this PR)

File: src/decision/location_selection.jl
Comment (arlowhite) [inline src/decision/location_selection.jl:303]: funny that it wants comma-space elsewhere, but not with types.

File: ext/AvizExt/AvizExt.jl
Comment (arlowhite) [reply ext/AvizExt/AvizExt.jl]: I've added `return nothing` in appropriate places. I think I've caught them all, but it won't hurt anything if I missed one since the formatter is preserving current behavior.

**Comment by Zapiano** [comment]
@arlowhite Now you can rebase and format all files. Thanks for waiting (and sorry it took so long).

**Comment by arlowhite** [comment]
I've disabled the `for in` change and `function` expansion. The documented defaults are false, but I think blue style turns them on.
* https://domluna.github.io/JuliaFormatter.jl/stable/#conditional_to_if
* https://domluna.github.io/JuliaFormatter.jl/stable/#short_to_long_function_def
* https://domluna.github.io/JuliaFormatter.jl/stable/#always_for_in

**Comment by arlowhite** [comment]
> Not a fan of these forced replacement of `in` to `∈` but I guess it's okay...
> 
> ```julia
> for r ∈ 1:n_rows
>         for c ∈ 1:n_cols
> ```

I've disabled it.

**Comment by arlowhite** [comment]
Only one place in code doesn't use for `in`
src\io\rme_result_io.jl
`location_centroids = [centroid(multipoly) for multipoly ∈ geodata.geom]`

So we might want to do
```
always_for_in=true
for_in_replacement="in"
```

File: ext/AvizExt/viz/taxa_dynamics.jl
Comment (Zapiano) [inline ext/AvizExt/viz/taxa_dynamics.jl:202]: Same here, probably if/else would be better.

File: src/ExtInterface/ReefMod/RMEDomain.jl
Comment (Zapiano) [inline src/ExtInterface/ReefMod/RMEDomain.jl:192]:
```suggestion
        T=Float64, 
        timesteps=timeframe_range, 
        locs=loc_ids, 
        species=functional_groups,
```

File: src/ExtInterface/ReefMod/ReefModDomain.jl
Comment (Zapiano) [inline src/ExtInterface/ReefMod/ReefModDomain.jl:223]: I'm not against it but this a little weird... maybe 

```julia
mean_init_icc = dropdims(mean(init_cc_per_taxa; dims=:scenario); dims=:scenario)
icc_data = (mean_init_icc ./ 100).data
```

File: src/analysis/rule_extraction.jl
Comment (Zapiano) [inline src/analysis/rule_extraction.jl:96]: What about

```
    condition(rule) = [
        c[2] == :L ? "$(c[1]) < $(c[3])" : "$(c[1]) ≥ $(c[3])" for c in rule.condition
    ]
```

?

File: src/decision/Criteria/SeedCriteria.jl
Comment (Zapiano) [inline src/decision/Criteria/SeedCriteria.jl:365]:
```suggestion
    cluster_frequency, 
    rule_violators_idx, 
    exceeded_clusters,
```

File: src/factors/Factors.jl
Comment (Zapiano) [inline src/factors/Factors.jl:56]:
```suggestion
        "continuous", 
        "ordered categorical", 
        "unordered categorical", 
        "ordered discrete",
```

File: src/io/rme_result_io.jl
Comment (Zapiano) [inline src/io/rme_result_io.jl:181]: I think a better way of doing this would be:

```
inputs.outplant_area_pct .== 0.0 .&&  inputs.enrichment_area_pct .== 0
```

I know, this is not just formatting. Maybe we can ignore this for now, but I just wanted to say that I really don't like this formatting (but again, I'm happy to ignore it).

File: src/io/rme_result_io.jl
Comment (Zapiano) [inline src/io/rme_result_io.jl:187]: similar comment as above

File: src/main_app.jl
Comment (Zapiano) [inline src/main_app.jl:117]:
```suggestion
        label="Guided median", 
        linecolor=:green, 
        alpha=0.4,
        xlabel="Year", 
        ylabel="Relative Cover",
        xticks=(1:75, year_axis))
```

File: src/main_app.jl
Comment (Zapiano) [inline src/main_app.jl:122]:
```suggestion
    p2 = plot(
        Y_ung[:mean] - Y_no[:mean]; 
        label="Guided - No Deployment (μ)",
        xlabel="Year", 
        ylabel="δ Relative Cover",
        xticks=(1:75, year_axis), 
        color=:red
    )
```

**Review by Zapiano — APPROVED** [review]
I left a lot of comments but, essentially, I don't mind if you ignore all of them. Whatever we decide, I'm happy to follow.

File: ext/AvizExt/AvizExt.jl
Comment (arlowhite) [reply ext/AvizExt/AvizExt.jl:135]: I don't think so. It's given as an argument. There's no current uses, but there is a commented one, which doesn't use a return value:
```
        # Clear loading animation
        # remove_loader(traj_display[2, 1], loader_anim_display)
```

File: src/main_app.jl
Comment (ConnectedSystems) [reply src/main_app.jl:122]: This `main_app` file is marked for removal. It was an earlier attempt to spin up a desktop GUI but isn't relevant for us any more.

File: ext/AvizExt/viz/rule_extraction.jl
Comment (arlowhite) [reply ext/AvizExt/viz/rule_extraction.jl:135]: up to you guys.
https://domluna.github.io/JuliaFormatter.jl/stable/#whitespace_ops_in_indices

Also, if we want whitespace in types 
https://domluna.github.io/JuliaFormatter.jl/stable/#whitespace_typedefs

Comment (ConnectedSystems) [reply ext/AvizExt/viz/rule_extraction.jl:135]: It's a config if I recall. I don't mind either way - sometimes with spaces is with more readable, other times without spaces is. Generally I prefer with spaces.

File: ext/AvizExt/viz/location_selection.jl
Comment (ConnectedSystems) [reply ext/AvizExt/viz/location_selection.jl:78]: I have a dissenting opinion.

Moving arguments around is something that's done during prototyping. Once prototyping is done, leaving the comma is untidy. If you have to adjust again (ideally months to years afterwards) then you just add the comma while prototyping.

But I think this discussion is moot because the point of having a formatting workflow is that you can do your thing and once you're finished the formatter standardises.

File: src/analysis/rule_extraction.jl
Comment (arlowhite) [reply src/analysis/rule_extraction.jl:96]: that's ok too, if I change to that, the formatter will leave it

File: src/decision/location_selection.jl
Comment (arlowhite) [reply src/decision/location_selection.jl:305]: This is a separate rule https://domluna.github.io/JuliaFormatter.jl/stable/#whitespace_typedefs

**Pull request merged**
